refactor: scripts, cross platform support

ferferga opened the pull request:
Needs testing on Mac. I did a minor adaptation to the Windows scripts in order to match the filenames used in the compose files and the new bash scripts, for consistency, but that's it, given I don't have much experience in batch scripting.

Key features:
* Use source to run the scripts instead of launching a subprocess, so all the context is carried across (variables, functions, etc)
* Use env so POSIX finds the shell in a cross-platform way between Linux distros and Macos (needs testing in Macos).
* Don't assume the only architecture is arm: if the current platform (queried by docker info) is not x86_64, emulation from buildx will always be needed!
* Detect the need of emulation inside the script, don't rely on the user
* Use Docker to generate the bcrypt password. We can assume docker is installed by the user since it's needed for running everything else, but we can't do the same for node.
* Remove all unnecessary duplication.

I added extra comments inside individual files. Marking as draft until tested in MacOS.

ferferga committed:
refactor: scripts, cross platform support

Signed-off-by: GitHub <noreply@github.com>

ferferga commented on README.md:
````diff
@@ -64,17 +64,8 @@ Before starting with the installation it is necessary to install or upgrade the
 
    ```bash
-   ./setup_infrastructure.sh
+   ./setup.sh
````
Given this repository is already called "infrastructure", I think it's implied that what we're setting up is the infrastructure, hence setup is enough.

alvarobernal2412 replied:
Thank!

ferferga commented on bin/initial_data.sh:
```diff
@@ -1,9 +1,6 @@
-#!/bin/sh
-
-username=$(sed -n '1p' /tmp/user_data)
-password=$(sed -n '2p' /tmp/user_data)
-email=$(sed -n '3p' /tmp/user_data)
```
These are now passed from the other script thanks to sourcing

alvarobernal2412 replied:
It is a pleasure to learn from your code improvements! Thank you!

ferferga commented on bin/initial_data.sh:
```diff
@@ -1,9 +1,6 @@
-email=$(sed -n '3p' /tmp/user_data)
+#!/usr/bin/env bash
 
+set -e
```
This tells bash that as soon as an error happens the script must be aborted immediatelly, removing the need of the ``execute_scripts`` function that we previously had.

alvarobernal2412 replied:
Cool!

ferferga commented on bin/prepare_environment.sh:
```diff
@@ -31,26 +33,26 @@ done
 
 ## If a folder is not created before doing a bind mount in Docker, the folder will be created with root permissions only.
-mkdir -p ../node-red-status
+mkdir -p node-red-status mysql 
```
Without creating this folder, further setups of the infrastructure would fail since the files couldn't be removed above (line 27)

alvarobernal2412 replied:
Thank you for noticing!

ferferga commented on bin/prepare_environment.sh:
```diff
@@ -61,22 +63,17 @@ echo
-hashed_password=$(node -e "console.log(require('bcrypt').hashSync(process.argv[1], 10));" "$password")
-escaped_hashed_password=$(printf '%s\n' "$hashed_password" | sed -e 's/[\/&]/\\&/g')
+# Hash the password
+docker pull epicsoft/bcrypt
```
This image is just 3 MB so this will be much faster than before as well

alvarobernal2412 commented on bin/prepare_environment.sh:
```diff
@@ -61,22 +63,14 @@ echo
+docker rmi epicsoft/bcrypt
 
 ## Replace the example_user and example_pass strings with the new user and password
 
```
There is a bug with the position of setting.js, it should be fixed by removing `../ before the filename

alvarobernal2412 replied:
I'm referring to this part of the code, sorry
<img width="645" alt="Captura de pantalla 2024-07-09 a las 16 40 42" src="https://github.com/statuscompliance/infrastructure/assets/80347035/f6fd1441-537a-4456-971a-4f82f3bfd1bf">

ferferga replied:
@alvarobernal2412 Good catch! Pushed the fix now!

alvarobernal2412 reviewed (changes requested):
Thank you so much for your contributions, it's a real pleasure to work with you!

ferferga committed:
fix: correct settings.js path

Signed-off-by: GitHub <noreply@github.com>

alvarobernal2412 commented on bin/cross_platform_utils.sh:
```diff
@@ -0,0 +1,35 @@
+
+function getDockerCompose() {
+  if needs_emulation; then
+    echo "docker-compose-emulation.yml"
```
For it to work properly you must rename it to `docker-compose-emulated.yml`.

alvarobernal2412 commented on start.sh:
```diff
@@ -0,0 +1,9 @@
+
+setupCrossPlatformEnvironment
+
+docker compose -f $(getDockerCompose) --env-file ../.env up -d --wait
```
For it to work properly you must change the `.env` location to `.` instead of `../` .

alvarobernal2412 commented on stop.sh:
```diff
@@ -0,0 +1,7 @@
+
+source bin/cross_platform_utils.sh
+
+docker compose -f $(getDockerCompose) --env-file ../.env down
```
For it to work properly you must change the `.env` location to `.` instead of `../` .

alvarobernal2412 reviewed (changes requested):
While testing the new changes, I found some minor changes that you should fix. After fixing it locally, I found the following problem.

![Captura de pantalla 2024-07-10 a las 10 16 54](https://github.com/statuscompliance/infrastructure/assets/80347035/7a7453e0-2bcd-4592-8280-46588ce6373e)

I tried to repeat the docker login and checked the dockerfiles but found nothing. Could it be from testing it in a fork?

ferferga committed:
fix: GNU sed not behaving consistently, login warnings from compose, print compose file used

Signed-off-by: GitHub <noreply@github.com>

ferferga committed:
refactor(docker): real healthchecks

Signed-off-by: GitHub <noreply@github.com>

ferferga committed:
refactor: use volume for mysql

Signed-off-by: GitHub <noreply@github.com>

ferferga commented on bin/initial_data.sh:
```diff
@@ -1,30 +1,22 @@
 echo "_______________________INITIAL DATA_______________________"
 
-## Wait for the server to be ready
-sleep 10
```
Unnecessary after the healthchecks

ferferga commented on bin/initial_data.sh:
```diff
@@ -1,30 +1,22 @@
 
 ## Create a new user
-curl -X POST http://localhost:3001/api/user/signUp \
+curl -s -X POST http://localhost:3001/api/user/signUp \
```
`-s` makes curl be silent and not output the progress

ferferga commented on bin/prepare_environment.sh:
```diff
@@ -16,41 +18,44 @@ else
+
 ## Clean up previous installations
-directories=("status-backend" "status-frontend" "node-red-status" "mysql" "collector-events")
+directories=("status-backend" "status-frontend" "node-red-status" "collector-events" ".env")
```
.env file is also part of our script and it was missing here, so we remove it as well during setup now.

ferferga commented on docker-compose-emulated.yml:
```diff
@@ -13,10 +12,12 @@ services:
     depends_on:
       mysql:
         condition: service_healthy
-    image: status-backend
```
This is the origin of the warnings in docker compose

ferferga commented on docker-compose-emulated.yml:
```diff
@@ -263,8 +263,6 @@ services:
     restart: always
     image: mysql:8.3.0
     container_name: mysql
-    expose:
-      - "3306"
```
Nowadays this is mostly just for documentation purposes (so it appears when we do ``docker inspect``), so docker compose add the tag to the container/image built during up. Useless at runtime.

ferferga reviewed (commented):
@alvarobernal2412 Really good catches, thank you!

Key things with these new commits:

* Found an inconsistency between GNU's sed (the one usually shipped with Linux distros) and MacOS (which is sourced from BSD iirc), fixed it so it works great no matter the version
* Added real healthchecks
* Style fixes, between console text there are some newlines instead of a huge wall of text
* Moved MySQL to a Docker volume (please let me know if this is unwanted). Due to the way the mysql container is built, even if we create the folder beforehand, the container wil take ownership of it. We could fix this by using ``sudo rm``. but escalating privileges in a script like this and for this reason doesn't seem the most sensible solution. It's strange that we ever need to deal with MySQL files anyway (unlike node-red, where I guess you guys work with mashups and so on), so I think they can be "abstracted" away from the user easily.

The reason why there were some warnings in compose was because you specified some containers to built, but also passed an imagen. Docker tries to fetch the image first from DockerHub, if it doesn't find it, then it fallbacks to the build process. In our case, since we want to use the local image always, I just removed the ``image`` key from the definition.
If in the future we want to publish the images, then the warnings will disappear since Docker will be able to find them in the registry (the warning it's just because it couldn't find them!)

alvarobernal2412 commented on bin/prepare_environment.sh:
```diff
@@ -16,41 +18,44 @@ else
+git clone -b feat/9-Manage_AI_requests https://github.com/statuscompliance/status-backend.git status-backend
 
-git clone -b feature/24-Bluejay_integration https://github.com/statuscompliance/status-frontend.git ../status-frontend
+git clone -b feature/24-Bluejay_integration https://github.com/statuscompliance/status-frontend.git status-frontend
```
To deploy status-frontend with nginx, you need to use the last frontend version (located in `feature/22-Create_initial_chatbot` branch).

ferferga replied:
Updated 👍

alvarobernal2412 commented on bin/prepare_environment.sh:
```diff
@@ -61,22 +66,23 @@ echo
 ## Replace the example_user and example_pass strings with the new user and password
+## Due to differences in GNU and Mac implementations, we don't do it in place
+
+function setVariables() {
```
This function creates a `settings.js-e` because of the `-i` flag. Remove this flag and it will probably work.

ferferga replied:
Ooops, good catch! 🚀

alvarobernal2412 reviewed (changes requested):
It won't be long now!

ferferga committed:
refactor: address review comments

Signed-off-by: GitHub <noreply@github.com>

ferferga committed:
refactor: speedup healthchecks

Signed-off-by: GitHub <noreply@github.com>

ferferga committed:
fix: all windows scripting

Windows scripting was not working at all. Adapted following the same
conventions we did for MacOS/Linux and upgraded to PowerShell instead,
which is the modern scripting language that all new scripts should use
in Windows

Signed-off-by: GitHub <noreply@github.com

alvarobernal2412 reviewed (approved):
The installation scripts for MacOS work correctly, so I accept this PR. However, it is important to test the Windows part in order to merge it.
